File: leetcode/mostBooked.cpp
/**********************************************
  > File Name		: mostBooked.cpp
  > Author		    : lunar
  > Email			: lunar_ubuntu@qq.com
  > Created Time	: Sun Sep  4 11:49:20 2022
  > Location        : Shanghai
  > Copyright@ https://github.com/xiaoqixian
 **********************************************/

#include <vector>
#include <algorithm>
#include <queue>
using namespace std;

class Solution {
private:
    struct Room {
        long long start, end;
        int number;
    };
public:
    int mostBooked(const int n, vector<vector<int>>& meetings) {
        sort(meetings.begin(), meetings.end(), [&](const vector<int>& v1, const vector<int>& v2) {return v1[0] >= v2[0];});

        auto cmp = [](const Room& v1, const Room& v2) {
            if (v1.end == v2.end) {
                return v1.number > v2.number;
            } else return v1.end > v2.end;
        };
        typedef priority_queue<Room, vector<Room>, decltype(cmp)> pq_type;

        pq_type pq(cmp);

        for (int i = 0; i < n; i++) {
            pq.push(Room {0, 0, i});
        }

        vector<int> use_times(n, 0);
        long long clock = 0;

        while (!meetings.empty()) {
            const vector<int> mt = meetings.back();
            meetings.pop_back();

            if (clock < mt[0]) {
                clock = mt[0];
            }

            vector<Room> buffer;
            int index = -1;

            while (!pq.empty()) {
                const auto& r = pq.top();
                if (r.end <= clock) {
                    buffer.emplace_back(pq.top());
                    if (index == -1 || r.number < buffer[index].number) {
                        index = buffer.size()-1;
                    }
                    pq.pop();
                } else break;
            }

            for (int i = 0; i < buffer.size(); i++) {
                if (i == index) {
                    continue;
                }
                pq.push(std::move(buffer.at(i)));
            }

            Room room;
            if (index == -1) {
                room = pq.top();
                pq.pop();
            } else {
                room = buffer[index];
            }


            if (clock < room.end) {
                clock = room.end;
            }

            room.start = clock;
            room.end = clock + (mt[1] - mt[0]);
            use_times[room.number]++;
            pq.push(std::move(room));
        }

        int res = -1, max_times = 0;
        for (int i = 0; i < n; i++) {
            if (use_times[i] > max_times) {
                max_times = use_times[i];
                res = i;
            }
        }
        return res;
    }
};

int main() {
    vector<vector<int>> vv({{1,20},{2,10},{3,5},{4,9},{6,8}});
    printf("%d\n", Solution().mostBooked(3, vv));
}

File: leetcode/most_booked.rs
/**********************************************
  > File Name		: most_booked.rs
  > Author		    : lunar
  > Email			: lunar_ubuntu@qq.com
  > Created Time	: Sun Sep  4 17:16:40 2022
  > Location        : Shanghai
  > Copyright@ https://github.com/xiaoqixian
 **********************************************/

trait Comp<T> {
    fn comp(&self, p2: &T) -> bool; //return true if self < p2.
}

struct PriorityQueue<T: Copy+Comp<T>> {
    queue: Vec<T>,
    alloc: usize //number of elements in queue
}

impl<T> PriorityQueue<T> where T: Copy+Comp<T> {
    fn new() -> Self {
        Self {
            queue: Vec::new(),
            alloc: 0
        }
    }

    fn size(&self) -> usize {
        self.alloc
    }

    fn empty(&self) -> bool {
        self.alloc == 0
    }

    fn exch(&mut self, i: usize, j: usize) {
        let temp = self.queue[i];
        self.queue[i] = self.queue[j];
        self.queue[j] = temp;
    }

    fn swim(&mut self, mut k: usize) {
        while k > 1 && self.queue[k].comp(&self.queue[k/2]) {
            self.exch(k, k/2);
            k /= 2;
        }
    }

    fn sink(&mut self, mut k: usize) -> usize {
        let mut j: usize;
        let size = self.alloc;
        
        while 2*k <= size {
            j = 2*k;
            if j < size && self.queue[j+1].comp(&self.queue[j]) {
                j += 1;
            }
            if !self.queue[j].comp(&self.queue[k]) {
                break;
            }
            self.exch(j, k);
            k = j;
        }
        k
    }

    fn pop(&mut self) -> Option<T> {
        if self.alloc == 0 {
            return None;
        }
        
        let res = Some(self.queue[1]);
        self.exch(1, self.alloc);
        self.alloc -= 1;
        self.sink(1);
        res
    }

    fn push(&mut self, item: T) {
        if self.queue.len() == 0 {
            self.queue.push(item);//queue[0] is not used.
        }
        if self.alloc == self.queue.len()-1 {
            self.queue.push(item);
            self.alloc += 1;
        } else {
            self.alloc += 1;
            self.queue[self.alloc] = item;
        }
        self.swim(self.alloc);
    }
}

type Room = (u64, u64, i32);

impl Comp<Room> for Room {
    fn comp(&self, p2: &Room) -> bool {
        if self.1 == p2.1 {
            self.2 > p2.2
        } else {
            self.1 > p2.1
        }
    }
}

struct Solution;

impl Solution {
    pub fn most_booked(n: i32, mut meetings: Vec<Vec<i32>>) -> i32 {
        meetings.sort_by(|v1, v2| {v1[0] < v2[0]});

        
    }
}
